File: Spring Boot/monolith/src/main/java/pl/edu/pg/benchmarking/monolith/tsp/algorithm2/Route2Service.java
package pl.edu.pg.benchmarking.monolith.tsp.algorithm2;

import org.springframework.stereotype.Service;
import pl.edu.pg.benchmarking.monolith.distance.SphericalDistanceService;
import pl.edu.pg.benchmarking.monolith.place.Place;
import pl.edu.pg.benchmarking.monolith.place.PlaceService;
import pl.edu.pg.benchmarking.monolith.tsp.PlaceNotFoundException;

import java.util.Arrays;
import java.util.Collection;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class Route2Service {
    private final PlaceService placeService;
    private final SphericalDistanceService distanceService;

    public Route2Service(PlaceService placeService, SphericalDistanceService distanceService) {
        this.placeService = placeService;
        this.distanceService = distanceService;
    }

    private Double findDistance(Integer placeId1, Integer placeId2) throws PlaceNotFoundException {
        Place place1 = placeService.getPlace(placeId1).orElseThrow(PlaceNotFoundException::new);
        Place place2 = placeService.getPlace(placeId2).orElseThrow(PlaceNotFoundException::new);
        return distanceService.calculateDistanceBetweenPoints(
                place1.getPoint(),
                place2.getPoint());
    }

    public List<List<Integer>> computePermutationRecursive(Integer[] elements) {
        return computePermutationRecursive(elements.length, elements, new LinkedList<>());
    }

    public List<List<Integer>> computePermutationRecursive(int n, Integer[] elements, List<List<Integer>> listOfPermutations) {
        if (n == 1) listOfPermutations.add(Arrays.stream(elements).collect(Collectors.toList()));
        else {
            for(int i = 0; i < n-1; i++) {
                computePermutationRecursive(n - 1, elements, listOfPermutations);
                if(n % 2 == 0)  swap(elements, i, n-1);
                else            swap(elements, 0, n-1);
            }
            computePermutationRecursive(n - 1, elements, listOfPermutations);
        }
        return listOfPermutations;
    }

    private void swap(Integer[] input, int a, int b) {
        Integer tmp = input[a];
        input[a] = input[b];
        input[b] = tmp;
    }

    public List<Integer> computeRoute(Collection<Integer> places) throws PlaceNotFoundException {
        List<List<Integer>> listOfPermutations = computePermutationRecursive(Arrays.copyOfRange(places.toArray(new Integer[0]), 1, places.size()));

        listOfPermutations.forEach(permutationList -> {
            int indexOfFirstElement = places.iterator().next();
            permutationList.add(0, indexOfFirstElement);
            permutationList.add(indexOfFirstElement);
        });

        List<Integer> shortestWay = null;
        Double shortestDistance = null;

        for (List<Integer> singlePermutationList : listOfPermutations) {
            double distance = 0;

            for (int i = 0; i < singlePermutationList.size() - 1; i += 1) {
                int placeId1 = singlePermutationList.get(i);
                int placeId2 = singlePermutationList.get(i + 1);
                distance += findDistance(placeId1, placeId2);
            }

            if (shortestDistance == null || shortestDistance.compareTo(distance) > 0) {
                shortestDistance = distance;
                shortestWay = singlePermutationList;
            }
        }
        return shortestWay;
    }
}

File: .NET/Benchmark/Services/Route2Service.cs
using Benchmark.Repositories;

namespace Benchmark.Services
{
    public class Route2Service
    {
        private readonly SphericalDistanceService _distanceService;
        private readonly PlaceRepository _placeRepository;

        public Route2Service(SphericalDistanceService distanceService, PlaceRepository placeRepository)
        {
            _distanceService = distanceService;
            _placeRepository = placeRepository;
        }

        public double FindDistance(int place1Id, int place2Id)
        {
            var cords1 = _placeRepository.FindById(place1Id);
            var cords2 = _placeRepository.FindById(place2Id);

            return _distanceService.CalculateDistance(cords1.GetPoint(), cords2.GetPoint());
        }

        public List<List<int>> ComputePermutationRecursive(int[] elements)
            => ComputePermutationRecursive(elements.Length, elements, new());

        public List<List<int>> ComputePermutationRecursive(int n, int[] elements, List<List<int>> listOfPermutations)
        {
            if (n == 1)
            {
                listOfPermutations.Add(elements.ToList());
            }

            else
            {
                for (var i = 0; i < n - 1; i++)
                {
                    ComputePermutationRecursive(n - 1, elements, listOfPermutations);
                    if (n % 2 == 0)
                        Swap(elements, i, n - 1);

                    else
                        Swap(elements, 0, n - 1);
                }

                ComputePermutationRecursive(n - 1, elements, listOfPermutations);
            }

            return listOfPermutations;
        }

        public List<int> ComputeRoute(List<int> places)
        {
            //Created array - places[1] ... places[length - 1], is this the thing which java is doing?
            var listOfPermutations = ComputePermutationRecursive(places.Skip(1).ToArray());

            foreach (var permutationList in listOfPermutations)
            {
                var firstElem = places[0];
                permutationList.Insert(0, firstElem);
                permutationList.Add(firstElem);
            }

            List<int> shortestWay = null;
            double? shortestDistance = null;

            foreach (var singlePermutationList in listOfPermutations)
            {
                double distance = 0;
                for (var i = 0; i < singlePermutationList.Count - 1; i++)
                {
                    var place1Id = singlePermutationList[i];
                    var place2Id = singlePermutationList[i + 1];

                    distance += FindDistance(place1Id, place2Id);
                }

                if (shortestDistance == null || shortestDistance?.CompareTo(distance) > 0)
                {
                    shortestDistance = distance;
                    shortestWay = singlePermutationList;
                }
            }

            return shortestWay;
        }

        #region Helpers

        private void Swap(int[] arr, int idx1, int idx2)
        {
            var tmp = arr[idx1];
            arr[idx1] = arr[idx2];
            arr[idx2] = tmp;
        }

        #endregion
    }
}
